### ISSUE TITLE
backpropGradient for ComputationGraph looking for 2dLabels

### ISSUE THREAD

**Issue by Joshuaalbert** [deeplearning4j/deeplearning4j#2133]
I have a Computation graph with 3d input and two RnnOutputLayer's. One is a regression and the other is a softmax. I've attached MSE and MXCENT loss functions to them, but I want to use the backpropGradient to set my own error, so the loss functions shouldn't matter(?). I'm getting a stacktrace
```
at org.deeplearning4j.nn.layers.recurrent.RnnOutputLayer.getLabels2d(RnnOutputLayer.java:128)
	at org.deeplearning4j.nn.layers.BaseOutputLayer.getGradientsAndDelta(BaseOutputLayer.java:182)
	at org.deeplearning4j.nn.layers.BaseOutputLayer.backpropGradient(BaseOutputLayer.java:161)
	at org.deeplearning4j.nn.layers.recurrent.RnnOutputLayer.backpropGradient(RnnOutputLayer.java:74)
	at org.deeplearning4j.nn.graph.vertex.impl.LayerVertex.doBackward(LayerVertex.java:114)
	at org.deeplearning4j.nn.graph.ComputationGraph.calcBackpropGradients(ComputationGraph.java:1088)
	at org.deeplearning4j.nn.graph.ComputationGraph.backpropGradient(ComputationGraph.java:1049)
``` 
which is failing at `if(labels.rank()==3) return reshape3dTo2d(labels);`
I shouldn't need to set the labels or input to propagate the errors back, so I'm not sure why it's looking for labels. 
---- Update ----
After looking around I see that backpropGradient in https://github.com/deeplearning4j/deeplearning4j/blob/master/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/layers/BaseOutputLayer.java doesn't use the epsilon in the argument at all, which is what we're trying to propagate! Instead it tries to compute the difference between labels and activation(preOut) in getGradientsAndDelta. 
```
public Pair<Gradient,INDArray> backpropGradient(INDArray epsilon) {
        Triple<Gradient,INDArray,INDArray> triple = getGradientsAndDelta(preOutput2d(true));	//Returns Gradient and delta^(this), not Gradient and epsilon^(this-1)
        INDArray delta = triple.getSecond();

        INDArray epsilonNext = params.get(DefaultParamInitializer.WEIGHT_KEY).mmul(delta.transpose()).transpose();
        return new Pair<>(triple.getFirst(),epsilonNext);
    }
```

**Comment by AlexDBlack** [comment]
Can you post your full model configuration?
As for "I want to use the backpropGradient to set my own error" that's probably the source of your problem. Sounds like you are trying to mix passing in external errors with output layers - it's one or the other, not both. Output layers calculate the error with respect to the labels.

**Comment by Joshuaalbert** [comment]
@AlexDBlack This is the model. Essentially it is a stack of dense relu which splits and goes to 1) stack of lstm 2) another relu stack and 3) another relu stack. Each layer of the lstm stack also feeds into the front of stacks 2 and 3. The outputs are at the end of stacks 2 and 3 which I set as rnn output layers. 

I didn't know I should not use output layers. 
**A)** How should I define them? 
**B)** If I never call setOutputs, and backprop goes back in reverse topological order (so it sets epsilons at the end and just rewinds the clock) I should be able to backprop from any single vertex? But then in that case, how do I do this? I thought the output determine where you inject the external error before rewinding?
```
conf = new Configuration(numInputs);
		
		ComputationGraphConfiguration.GraphBuilder confBuilder = new NeuralNetConfiguration.Builder()
				.seed(Constants.NEURAL_NET_SEED)
                .iterations(1)
                .optimizationAlgo(OptimizationAlgorithm.STOCHASTIC_GRADIENT_DESCENT)
                .learningRate(conf.learningRate)
                //.updater(Updater.NESTEROVS).momentum(0.9)
                //.updater(Updater.RMSPROP).rmsDecay(conf.getRmsDecay())
                .updater(Updater.ADAM)
                .weightInit(WeightInit.XAVIER)
                .regularization(true)
                .l2(conf.l2)
                .graphBuilder()
                .setInputTypes(InputType.recurrent(numInputs))
                //.setInputTypes(InputType.feedForward(numInputs))
                .addInputs("input")//.addVertex("input2prefilter", new LastTimeStepVertex("input"), "input")
                .addLayer("prefilter0", new DenseLayer.Builder()
                        .nIn(numInputs)
                        .nOut(conf.numPrefilterNodes)
                        .activation("relu")
                        .build(), "input");

		//build prefilter
		for (int i = 1; i < conf.numPrefilterLayers; i++) {
            confBuilder.addLayer("prefilter"+i, new DenseLayer.Builder()
            		.nIn(conf.numPrefilterNodes)
	    			.nOut(conf.numPrefilterNodes)
	    			.activation("relu")
	    			.build(), "prefilter"+(i - 1));
        }
		
		//merge 2d to 3d
		confBuilder.addVertex("prefilter2controller", new DuplicateToTimeSeriesVertex("input"), "prefilter"+(conf.numPrefilterLayers - 1));
		//build controller
		String[] controllerLayerLabels = new String[conf.numControllerLayers];
		confBuilder.addLayer("controller0", new GravesLSTM.Builder()
        		.nIn(conf.numPrefilterNodes)
    			.nOut(conf.numControllerNodes)
    			.activation("tanh")
    			.build(), "prefilter2controller");
		controllerLayerLabels[0] = "controller0";
		for (int i = 1; i < conf.numControllerLayers; i++) {
			confBuilder.addLayer("controller"+i, new GravesLSTM.Builder()
	        		.nIn(conf.numControllerNodes)
	    			.nOut(conf.numControllerNodes)
	    			.activation("tanh")
	    			.build(), "controller"+(i-1));
			controllerLayerLabels[i] = "controller"+i;
        }
		//merge output from each controller layer and then to 2d
		confBuilder.addVertex("controllerMerge", new MergeVertex(), controllerLayerLabels);
		confBuilder.addVertex("controller2d", new LastTimeStepVertex("input"), "controllerMerge");
	
		//build precritic
		confBuilder.addLayer("precritic0", new DenseLayer.Builder()
                .nIn(conf.numControllerNodes*conf.numControllerLayers+conf.numPrecriticNodes)
                .nOut(conf.numPrecriticNodes)
                .activation("relu")
                .build(), "controller2d","prefilter"+(conf.numPrefilterLayers-1));
		for (int i = 1; i < conf.numPrecriticLayers; i++) {
			confBuilder.addLayer("precritic"+i, new DenseLayer.Builder()
	                .nIn(conf.numPrecriticNodes)
	                .nOut(conf.numPrecriticNodes)
	                .activation("relu")
	                .build(), "precritic"+(i - 1));
		}
		//build preactor
		confBuilder.addLayer("preactor0", new DenseLayer.Builder()
                .nIn(conf.numControllerNodes*conf.numControllerLayers+conf.numPreactorNodes)
                .nOut(conf.numPrecriticNodes)
                .activation("relu")
                .build(), "controller2d","prefilter"+(conf.numPrefilterLayers-1));
		for (int i = 1; i < conf.numPreactorLayers; i++) {
			confBuilder.addLayer("preactor"+i, new DenseLayer.Builder()
	        		.nIn(conf.numPreactorNodes)
	    			.nOut(conf.numPreactorNodes)
	    			.activation("relu")
	    			.build(), "preactor"+(i - 1));
		}
		
		//define outputs
		
		confBuilder.addLayer("value", new RnnOutputLayer.Builder(LossFunctions.LossFunction.MSE)
                .activation("identity").nIn(conf.numPrecriticNodes)
                .nOut(1).build(), "precritic"+(conf.numPrecriticLayers-1));
		confBuilder.addLayer("softmax", new RnnOutputLayer.Builder(LossFunctions.LossFunction.MCXENT)
		        .activation("softmax").nIn(conf.numPreactorNodes)
		        .nOut(numActions).build(), "preactor"+(conf.numPreactorLayers-1));
		confBuilder.setOutputs("value", "softmax");


		ComputationGraphConfiguration cgconf = confBuilder.pretrain(false).backprop(true).build();
		ComputationGraph model = new ComputationGraph(cgconf);
		model.init();
		model.setListeners(new ScoreIterationListener(Constants.NEURAL_NET_ITERATION_LISTENER));
		for (Layer layer : model.getLayers()){
			System.out.println(layer.input());
		}
		
		return model;
```

**Comment by Joshuaalbert** [comment]
solved
